### PR TITLE
Using assistant in AutomaticSpeechRecognitionPipeline with different encoder size

### PR DESCRIPTION
**Pull request by kamilakesbi**
This PR aims at fixing issue #30611: 

- First: an error will be thrown if the assistant and main models encoders don't have the same size, and the assistant is loaded using `AutoModelForCausalLM`. 

- Second: This PR makes the pipeline work when using an assistant with a different encoder size (loaded with   `AutoModelForSpeechSeq2Seq`) than the main model: 

When using AutomaticSpeechRecognitionPipeline, If we use an assistant with a different encoder size than the main model , the pipeline is broken and we get the following error message: 

```
ValueError: Whisper expects the mel input features to be of length 3000, but found 1500. Make sure to pad the input mel features to 3000.
```

## Explanation of the solution

When doing short form generation with the pipeline, `input_features` aren't passed to the generate method, which instead takes the output of the main model's encoder. 

If the main model and the assistant don't share the same encoder, the encoder_output passed to generate cannot be used by the assistant for generation, and we get an error. 

The solution here is to also pass the `input_features` to the generate method to be used by the assistant. 

## Who can review?

@sanchit-gandhi

### CODE REVIEW

File: src/transformers/models/whisper/generation_whisper.py [hunk 1]
```diff
@@ -474,6 +474,15 @@ def generate(
                 "The input name `inputs` is deprecated. Please make sure to use `input_features` instead.",
                 FutureWarning,
             )
+
+        if "assistant_model" in kwargs:
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/generation_whisper.py:478]: This check should live in `generate`, as it's relevant to all encoder-decoder models. Furthermore, it should be checking that:
1. If the main model is `AutoModelForSpeechSeq2Seq` (i.e. encoder-decoder) and the assistant model is `AutoModelForCausalLM` (i.e. decoder-only), then the encoder dims have to match (as well as the tokeniser's). Ideally, we'd check for [all `encoder_xyz` attributes](https://huggingface.co/openai/whisper-large-v3/blob/1ecca609f9a5ae2cd97a576a9725bc714c022a93/config.json#L23-L26), not just `encoder_attention_heads`
2. If the main model is `AutoModelForSpeechSeq2Seq` (i.e. encoder-decoder) and the assistant model is also `AutoModelForSpeechSeq2Seq` (i.e. encoder-decoder), then the encoder dims can be different. The only criteria here is that both model use the same tokenizer (we can do a shallow check by looking at the vocab size and making sure they're the same)

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 2]
```diff
@@ -493,6 +493,9 @@ def _forward(self, model_inputs, return_timestamps=False, **generate_kwargs):
             else:
                 generate_kwargs["encoder_outputs"] = encoder(inputs, attention_mask=attention_mask)
 
+            if "assistant_model" in generate_kwargs:
+                generate_kwargs["input_features"] = inputs
```
Comment (sanchit-gandhi) [inline src/transformers/pipelines/automatic_speech_recognition.py:497]: This logic should be under the `if` statement for encoder-decoder models only? i.e.
```python
if self.type in {"seq2seq", "seq2seq_whisper"}:
    ...
    if "input_features" in model_inputs:
        ...
    elif "input_values" in model_inputs:
        ...
    else:
        ...
    if "assistant_model" in generate_kwargs:
        generate_kwargs["input_features"] = inputs
    ...

Comment (sanchit-gandhi) [reply src/transformers/pipelines/automatic_speech_recognition.py:497]: It would be a good exercise to figure out _why_ making this change works as well - if you could then explain on the PR description that would be grand

**Review by sanchit-gandhi — COMMENTED** [review]
Thanks for taking this up @kamilakesbi - left some comments below!

File: src/transformers/models/whisper/generation_whisper.py (hunk 1 above)
Comment (kamilakesbi) [reply src/transformers/models/whisper/generation_whisper.py:478]: Would be better, but as far as I know, those are the conditions for audio models no ? shouldn't we also generalize these observations to non-audio models compatible with assistants ?

File: src/transformers/generation/utils.py [hunk 3]
```diff
@@ -1087,6 +1087,24 @@ def _validate_model_class(self):
                 exception_message += f" Please use one of the following classes instead: {generate_compatible_classes}"
             raise TypeError(exception_message)
 
+    def _validate_assistant(self, assistant_model):
+        if assistant_model is not None:
+            if type(self).__name__ in MODEL_FOR_SPEECH_SEQ_2_SEQ_MAPPING._model_mapping.values():
+                if type(assistant_model).__name__ in MODEL_FOR_CAUSAL_LM_MAPPING._model_mapping.values():
```
Comment (sanchit-gandhi) [inline src/transformers/generation/utils.py:1093]: Would it be more general if we could simply check: `self.config.is_encoder_decoder` and `assistant_model.config.is_encoder_decoder`? If both True, then we know both are enc-dec models. If True and False, then the assistant is dec-only.

**Review by sanchit-gandhi — COMMENTED** [review]
Small comment regarding the generality of the check (note that in `generation/utils.py`, we are assuming that all checks + functionality can be applied to **all** models in the library that are generate-compatible, not just speech recognition ones)

**Comment by sanchit-gandhi** [comment]
Is there a test that confirms correctness after the fix? There's likely a relevant slow pipeline test that was either failing, or was not rigorous enough

**Comment by gante** [comment]
Please have a look at #30726 for an alternative fix -- IMO, the root source of problems is in the ASR pipeline, and not in generate :)

File: src/transformers/generation/utils.py (hunk 3 above)
Comment (kamilakesbi) [reply src/transformers/generation/utils.py:1093]: Could it be that the assistant is encoder only ? in this case it would be perhaps better to check that `assistant_model.config.is_decoder`. 

However, when using whisper models with `AutoModelForCausalLM`, `self.config.is_decoder` is False. Is that expected ?

Comment (kamilakesbi) [reply src/transformers/generation/utils.py:1093]: I can make the changes with `self.config.is_encoder_decoder` and `assistant_model.config.is_encoder_decoder` for now.

File: tests/pipelines/test_pipelines_automatic_speech_recognition.py [hunk 4]
```diff
@@ -1007,6 +1010,119 @@ def test_whisper_language(self):
...
+        assert transcription_ass == transcription_non_ass
+        assert (
+            transcription_ass
+            == " Mr. Quilter is the apostle of the middle classes and we are glad to welcome his gospel."
+        )
+
+        assert total_time_non_assist > total_time_assist, "Make sure that assistant decoding is faster"
+
+    @slow
+    def test_speculative_decoding_whisper_distil(self):
```
Comment (sanchit-gandhi) [inline tests/pipelines/test_pipelines_automatic_speech_recognition.py:1071]: Feel free to propagate the above changes here as well

**Review by sanchit-gandhi — APPROVED** [review]
Pipeline changes LGTM, just some minor suggestions regarding the slow tests. Would love a second opinion from generate expert @gante on the assistant model validation!

File: tests/pipelines/test_pipelines_automatic_speech_recognition.py (hunk 4 above)
Comment (kamilakesbi) [reply tests/pipelines/test_pipelines_automatic_speech_recognition.py:1059]: Done!

**Comment by kamilakesbi** [comment]
I think this PR is ready to be merged! 

cc @amyeroberts @gante if you want to have a look ;)

File: src/transformers/generation/utils.py [hunk 5]
```diff
@@ -1087,6 +1087,24 @@ def _validate_model_class(self):
...
+                        for attr in attributes_to_check
+                    )
+                    if not are_equal:
+                        raise ValueError(
+                            "The main model and the assistant don't have encoders of the same size. "
+                            "Ensure you load the assistant with the correct encoder-decoder class, e.g. `AutoModelForSpeechSeq2Seq` for Whisper."
+                        )
+
+                if not self.config.vocab_size == assistant_model.config.vocab_size:
+                    raise ValueError("Make sure the main and assistant model use the same tokenizer")
```
Comment (amyeroberts) [inline src/transformers/generation/utils.py:1106]: Code should be indented as little as possible. We can instead do early returns, which more clearly separates the checking logic and keeps the main logic as unindented as possible

```suggestion
        if assistant_model is None or not self.config.is_encoder_decoder:
            return

        if not assistant_model.config.is_encoder_decoder:
            attributes_to_check = [attr for attr in dir(self.config) if attr.startswith("encoder_")]
            are_equal = all(
                getattr(self.config, attr) == getattr(assistant_model.config, attr)
                for attr in attributes_to_check
            )
            if not are_equal:
                raise ValueError(
                    "The main model and the assistant don't have encoders of the same size. "
                    "Ensure you load the assistant with the correct encoder-decoder class, e.g. `AutoModelForSpeechSeq2Seq` for Whisper."
                )

        if not self.config.vocab_size == assistant_model.config.vocab_size:
            raise ValueError("Make sure the main and assistant model use the same tokenizer")
```

File: src/transformers/generation/utils.py [hunk 6]
```diff
@@ -1087,6 +1087,24 @@ def _validate_model_class(self):
...
+                    attributes_to_check = [attr for attr in dir(self.config) if attr.startswith("encoder_")]
+                    are_equal = all(
+                        getattr(self.config, attr) == getattr(assistant_model.config, attr)
+                        for attr in attributes_to_check
+                    )
+                    if not are_equal:
+                        raise ValueError(
+                            "The main model and the assistant don't have encoders of the same size. "
+                            "Ensure you load the assistant with the correct encoder-decoder class, e.g. `AutoModelForSpeechSeq2Seq` for Whisper."
+                        )
```
Comment (amyeroberts) [inline src/transformers/generation/utils.py:1103]: There's a lot of assumptions happening in this check, which I don't think stand. For one, there might be properties which don't match e.g. `encoder_layerdrop` which have nothing to do with size, making this exception misleading. 

Is it necessary for the assistant to have exactly the same number of e.g. attention heads for all model types, even non-audio ones?

File: src/transformers/generation/utils.py [hunk 7]
```diff
@@ -1087,6 +1087,24 @@ def _validate_model_class(self):
                 exception_message += f" Please use one of the following classes instead: {generate_compatible_classes}"
             raise TypeError(exception_message)
 
+    def _validate_assistant(self, assistant_model):
+        if assistant_model is not None:
+            if self.config.is_encoder_decoder:
+                if not assistant_model.config.is_encoder_decoder:
```
Comment (amyeroberts) [inline src/transformers/generation/utils.py:1093]: If the assistant isn't an encoder-decoder model, then I wouldn't expect it to have `encoder_xxx` values in its config

File: tests/pipelines/test_pipelines_automatic_speech_recognition.py [hunk 8]
```diff
@@ -1007,6 +1010,95 @@ def test_whisper_language(self):
...
+        transcription_ass = pipe(sample)["text"]
+        total_time_non_assist = time.time() - start_time
+
+        assert transcription_ass == transcription_non_ass
+        assert (
+            transcription_ass
+            == " Mr. Quilter is the apostle of the middle classes and we are glad to welcome his gospel."
+        )
+
+        assert total_time_non_assist > total_time_assist, "Make sure that assistant decoding is faster"
```
Comment (amyeroberts) [inline tests/pipelines/test_pipelines_automatic_speech_recognition.py:1056]: Library convention is to use unittest for asserts 

```suggestion
        self.assertEqual(transcription_ass, transcription_non_ass)
        self.assertEqual(
            transcription_ass,
            " Mr. Quilter is the apostle of the middle classes and we are glad to welcome his gospel."
        )
        self.assertTrue(total_time_non_assist > total_time_assist, "Make sure that assistant decoding is faster") 
```

File: tests/pipelines/test_pipelines_automatic_speech_recognition.py [hunk 9]
```diff
@@ -1007,6 +1010,95 @@ def test_whisper_language(self):
...
+
+        start_time = time.time()
+        transcription_non_ass = pipe(sample.copy(), generate_kwargs={"assistant_model": assistant_model})["text"]
+        total_time_assist = time.time() - start_time
+
+        start_time = time.time()
+        transcription_ass = pipe(sample)["text"]
+        total_time_non_assist = time.time() - start_time
+
+        assert transcription_ass == transcription_non_ass
```
Comment (amyeroberts) [inline tests/pipelines/test_pipelines_automatic_speech_recognition.py:1095]: Same here re using unittest

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on this! 

I've just done a quick pass over and have some outstanding Qs. I'll review again once @gante has confirmed the update to the generation validation is OK

File: src/transformers/generation/utils.py (hunk 6 above)
Comment (sanchit-gandhi) [reply src/transformers/generation/utils.py]: There are two cases for assisted generation with an encoder-decoder model.

### Case 1

If the encoder of the main model matches that of the assistant model, we only need to load:
1. The main model as encoder-decoder
2. The assistant model as decoder-only

In doing so, we only have one copy of the encoder (which we know is the same for both models), and only have to run it forward once. This is done with the following `AutoModelForCausalLM` class:

```python
from transformers import AutoModelForCausalLM, AutoModelForSpeechSeq2Seq

model = AutoModelForSpeechSeq2Seq.from_pretrained("openai/whisper-large-v3")
assistant_model = AutoModelForCausalLM.from_pretrained("distil-whisper/distil-large-v3")

pred_ids = model.generate(..., assistant_model=assistant_model)
```

### Case 2

If the encoder's are **different** for the main and assistant models, both need to be loaded as encoder-decoder:

```python
from transformers import AutoModelForSpeechSeq2Seq

model = AutoModelForSpeechSeq2Seq.from_pretrained("openai/whisper-large-v2")
assistant_model = AutoModelForSpeechSeq2Seq.from_pretrained("openai/whisper-tiny")

pred_ids = model.generate(..., assistant_model=assistant_model)
```

### Relation to this PR

We were seeing issues where users would have main and assistant models with **different encoders**, but were loading the assistant model as **decoder-only** (e.g. in the code example #30407). This is the motivation for adding a check to ensure the encoder's match between the main and assistant models (cc @gante).

@amyeroberts is indeed correct that any regularisation parameters can differ, but all encoder dimensions have to match.

Note that this is still an imperfect check: we don't actually check that the pre-trained weights of the encoder's match for the main and assistant models, which is something we still assume. 

However, it's better than the silent failure users were seeing before.

File: src/transformers/generation/utils.py (hunk 7 above)
Comment (sanchit-gandhi) [reply src/transformers/generation/utils.py]: This is because we're only loading the decoder from an encoder-decoder model, see above comment: https://github.com/huggingface/transformers/pull/30637#discussion_r1601879851

File: src/transformers/generation/utils.py (hunk 6 above)
Comment (sanchit-gandhi) [reply src/transformers/generation/utils.py]: Would be interested in hearing what you think regarding the encoder checks given the above @gante @amyeroberts!

File: src/transformers/generation/utils.py [hunk 10]
```diff
@@ -1087,6 +1087,25 @@ def _validate_model_class(self):
                 exception_message += f" Please use one of the following classes instead: {generate_compatible_classes}"
             raise TypeError(exception_message)
 
+    def _validate_assistant(self, assistant_model):
+        if assistant_model is None or not self.config.is_encoder_decoder:
+            return
+
+        if not assistant_model.config.is_encoder_decoder:
```
Comment (gante) [inline src/transformers/generation/utils.py:1094]:
```suggestion
        if assistant_model is None:
            return

        if self.config.is_encoder_decoder and not assistant_model.config.is_encoder_decoder:
```
Suggestion for readability and future-proofing:
1. This function is called `_validate_assistant`, so let's make the quick escape route depend exclusively on whether `assistant_model` exists
2. By making the `if` clause this way, we know that these checks are relative to the specific case of encoder-decoder model with a decoder-only assistant 🤗

File: src/transformers/generation/utils.py [hunk 11]
```diff
@@ -1087,6 +1087,25 @@ def _validate_model_class(self):
...
+            attributes_to_check = ["encoder_attention_heads", "encoder_ffn_dim", "encoder_layers"]
+            attributes_to_check = [attr for attr in dir(self.config) if attr in attributes_to_check]
+            are_equal = all(
+                getattr(self.config, attr) == getattr(assistant_model.config, attr) for attr in attributes_to_check
+            )
+            if not are_equal:
+                raise ValueError(
+                    "The main model and the assistant don't have encoders of the same size. "
```
Comment (gante) [inline src/transformers/generation/utils.py:1102]:
```suggestion
                    "The main model and the assistant don't have compatible encoder-dependent input shapes. "
```
Suggestion: the assistant is decoder-only, so saying "...assistant don't have encoders of the same size" doesn't make much sense :D I'd reword it into something that points towards the encoder-dependent inputs.

**Review by gante — APPROVED** [review]
LGTM, thank you for fixing! 💪 I've added a few minor nits to help with readability.

(I see that you've included the diff from #30726, I'm going to close that PR :D)

File: src/transformers/generation/utils.py (hunk 10 above)
Comment (gante) [reply src/transformers/generation/utils.py:1094]: @sanchit-gandhi @amyeroberts I think this suggestion and the one below will help clear the questions like the ones on [this thread above](https://github.com/huggingface/transformers/pull/30637#discussion_r1601631205) :)

File: src/transformers/generation/utils.py [hunk 12]
```diff
@@ -1087,6 +1087,25 @@ def _validate_model_class(self):
                 exception_message += f" Please use one of the following classes instead: {generate_compatible_classes}"
             raise TypeError(exception_message)
 
+    def _validate_assistant(self, assistant_model):
+        if assistant_model is None or not self.config.is_encoder_decoder:
+            return
+
+        if not assistant_model.config.is_encoder_decoder:
+            attributes_to_check = ["encoder_attention_heads", "encoder_ffn_dim", "encoder_layers"]
+            attributes_to_check = [attr for attr in dir(self.config) if attr in attributes_to_check]
```
Comment (gante) [inline src/transformers/generation/utils.py:1096]:
```suggestion
            attributes_to_check = [attr for attr in dir(assistant_model.config) if attr in attributes_to_check]
```
It's okay if the main model has these attributes and the assistant doesn't, but not the other way around :) e.g. we can have an encoder-decoder model with a encoder-only model that has no encoder-related inputs

**Comment by amyeroberts** [comment]
@kamilakesbi It still needs a final core maintainer review and approval before merge (+ resolution of conflicts) :). I'll review now

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding and iterating on this! 

Only thing left to add are tests for `_validate_assistant` - there should be tests making sure that it correctly raises exceptions for the two cases it's checking for

**Comment by kamilakesbi** [comment]
hi @amyeroberts, I've added a slow test which pass :) I think quality checks fails are unrelated to this PR, and will be fixed by [#30932](https://github.com/huggingface/transformers/pull/30932).

File: tests/generation/test_utils.py [hunk 13]
```diff
@@ -2846,6 +2847,67 @@ def test_assisted_decoding_num_assistant_tokens_heuristic_transient_schedule(sel
         # update_candidate_strategy is called once but assistant_model.generation_config.num_assistant_tokens should stay 5
         self.assertEqual(assistant_model.generation_config.num_assistant_tokens, 5)
 
+    @slow
+    def test_validate_assistant(self):
```
Comment (amyeroberts) [inline tests/generation/test_utils.py:2851]: Nice!

**Review by amyeroberts — APPROVED** [review]
Thanks for adding the tests - looks great!

File: tests/generation/test_utils.py (hunk 13 above)
Comment (sanchit-gandhi) [reply tests/generation/test_utils.py:2851]: +1, nice test @kamilakesbi!

**Comment by sanchit-gandhi** [comment]
Quick checklist before merge:
- [ ] Resolve all comment threads that have been addressed
- [ ] Fix the merge conflict in `automatic_speech_recognition.py`
- [ ] Rebase onto main to get the style fixes from #30932 (once the PR is merged)
- [ ] Ping me to get this merged!

**Comment by sanchit-gandhi** [comment]
Nice work @kamilakesbi!
